pkg/cache/metrics_test.go

```diff
@@ -0,0 +1,97 @@
+//go:generate mockgen --build_flags=--mod=mod -package cache -destination ./mocks_test.go github.com/rancher/lasso/pkg/client SharedClientFactory
+package cache
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/rancher/lasso/pkg/client"
+	"github.com/rancher/lasso/pkg/metrics"
+	"go.uber.org/mock/gomock"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+)
+
+func setupMockSharedClientFactory(t *testing.T, cf *MockSharedClientFactory, gvr schema.GroupVersionResource, gvk schema.GroupVersionKind) {
+	t.Helper()
+
+	testClient := &client.Client{}
+	scheme := runtime.NewScheme()
+	if err := clientgoscheme.AddToScheme(scheme); err != nil {
+		t.Fatal(err)
+	}
+
+	cf.EXPECT().ResourceForGVK(gvk).DoAndReturn(func(gvk schema.GroupVersionKind) (schema.GroupVersionResource, bool, error) {
+		return gvr, true, nil
+	}).AnyTimes()
+	cf.EXPECT().ForKind(gvk).DoAndReturn(func(gvk schema.GroupVersionKind) (*client.Client, error) {
+		return testClient, nil
+	}).AnyTimes()
+	cf.EXPECT().NewObjects(gvk).DoAndReturn(func(gvk schema.GroupVersionKind) (runtime.Object, runtime.Object, error) {
+		obj, err := scheme.New(gvk)
+		if err != nil {
+			return nil, nil, err
+		}
+		objList, err := scheme.New(schema.GroupVersionKind{
+			Group:   gvk.Group,
+			Version: gvk.Version,
+			Kind:    gvk.Kind + "List",
+		})
+
+		return obj, objList, err
+	}).AnyTimes()
+	cf.EXPECT().ForResourceKind(gvr, gvk.Kind, true).DoAndReturn(func(gvr schema.GroupVersionResource, kind string, namespaced bool) *client.Client {
+		return testClient
+	})
+}
+
+func Test_sharedCacheFactory_metrics_collection(t *testing.T) {
+	cf := NewMockSharedClientFactory(gomock.NewController(t))
+	setupMockSharedClientFactory(t, cf, corev1.SchemeGroupVersion.WithResource("configmaps"), corev1.SchemeGroupVersion.WithKind("ConfigMap"))
+	setupMockSharedClientFactory(t, cf, rbacv1.SchemeGroupVersion.WithResource("roles"), rbacv1.SchemeGroupVersion.WithKind("Role"))
+
+	scf := NewSharedCachedFactory(cf, &SharedCacheFactoryOptions{MetricsCollectionPeriod: time.Millisecond * 200}).(*sharedCacheFactory)
+	if _, err := scf.ForKind(corev1.SchemeGroupVersion.WithKind("ConfigMap")); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := scf.ForKind(rbacv1.SchemeGroupVersion.WithKind("Role")); err != nil {
+		t.Fatal(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	ctx = metrics.WithContextID(ctx, "test-ctx")
+
+	// Start metrics collection:
+	// It will report 0 items for every kind, since indexers is registered but not started, which is enough for testing
```

Comment:
I think it would be worth it to verify that the total changes based on how many are in the store/indexer. Would you mind explaining why you don't think that's necessary?

Reply:
Just about here we could do something like this...

```go
    testCache, err := scf.ForResourceKind(corev1.SchemeGroupVersion.WithResource("configmaps"), "ConfigMap", true)
    if err != nil {
        t.Fatal(err)
    }
    testConfigMap := &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: "testing", Namespace: "default"}, TypeMeta: metav1.TypeMeta{Kind: "ConfigMap", APIVersion: "v1"}}
    if err := testCache.GetStore().Add(testConfigMap); err != nil {
        t.Fatal(err)
    }   
```

Which would result in the metric count being `1` for ConfigMaps.

This would test that we're recording the number of resources in the cache correctly...

You could then call `testCache.GetStore().Delete(testConfigMap)` and verify that the count goes back to zero before cancelling the context and triggering the removal of all metrics?

Reply:
Good point, thanks, added!